cabal-install/tests/UnitTests/Distribution/Client/Dependency/Modular/Solver.hs

```diff
@@ -85,9 +88,25 @@ tests = [
         , runTest $ mkTestLangs [Haskell98] dbLangs1 "unsupportedIndirect" ["B"] Nothing
         , runTest $ mkTestLangs [Haskell98, Haskell2010, UnknownLanguage "Haskell3000"] dbLangs1 "supportedUnknown" ["C"] (Just [("A",1),("B",1),("C",1)])
         ]
+
+     , testGroup "Soft Constraints" [
+          runTest $ soft [("A", mkvrThis [1,0,0])]      $ mkTest db13 "selectPreferredVersionSimple" ["A"] (Just [("A", 1)])
+        , runTest $ soft [("A", mkvrOrEarlier [2,0,0])] $ mkTest db13 "selectPreferredVersionSimple2" ["A"] (Just [("A", 2)])
+        , runTest $ soft [("A", mkvrOrEarlier [2,0,0])
+                         ,("A", mkvrOrEarlier [1,0,0])] $ mkTest db13 "selectPreferredVersionMultiple" ["A"] (Just [("A", 1)])
+        , runTest $ soft [("A", mkvrOrEarlier [1,0,0])
+                         ,("A", mkvrOrEarlier [2,0,0])] $ mkTest db13 "selectPreferredVersionMultiple2" ["A"] (Just [("A", 1)])
+        ]
     ]
   where
     indep test = test { testIndepGoals = True }
+    soft prefs test = test { testSoftConstraints = prefs }
+
+mkvrThis :: [Int] -> V.VersionRange
+mkvrThis = V.thisVersion . V.makeVersion 
+
+mkvrOrEarlier :: [Int] -> V.VersionRange
+mkvrOrEarlier = V.orEarlierVersion . V.makeVersion
```

Comment:
You could simplify the versions in the test cases by making `mkvrThis` and `mkvrOrEarlier` both have type `ExamplePkgVersion -> V.VersionRange`.